lib/DataLayoutAnalysis/Backend/DLAMakeModelTypes.cpp

```diff
@@ -34,82 +34,71 @@ using LTSN = LayoutTypeSystemNode;
 
 /// Create a single-field struct to wrap a given type
-static QualifiedType createStructWrapper(const LTSN *N,
-                                         const QualifiedType &T,
-                                         PtrFieldsMap &PointerFieldsToUpdate,
-                                         TupleTree<model::Binary> &Model,
-                                         uint64_t Offset = 0ULL,
-                                         uint64_t WrapperSize = 0ULL) {
+static model::UpcastableType
+createStructWrapper(const LTSN *N,
+                    UpcastableType &&T,
+                    PtrFieldsMap &PointerFieldsToUpdate,
+                    TupleTree<model::Binary> &Model,
+                    uint64_t Offset = 0ULL,
+                    uint64_t WrapperSize = 0ULL) {
   // Create struct
-  auto [Struct, Path] = Model->makeTypeDefinition<model::StructDefinition>();
+  auto [Struct, NewType] = Model->makeStructDefinition();
 
   // Create and insert field in struct
-  StructField Field{ Offset, {}, {}, {}, T };
-  const auto &[FieldIt, Inserted] = Struct.Fields().insert(Field);
-  revng_assert(Inserted);
-
-  // If the field has pointer type, save the corresponding qualified type for
-  // updating it later
-  if (T.isPointer()) {
-    bool New = PointerFieldsToUpdate[N].insert(&FieldIt->Type()).second;
+  model::StructField &Field = Struct.addField(Offset, std::move(T));
+
+  // If the field has pointer type, save the corresponding type, since we'll
+  // want to update it later.
+  if (Field.Type()->isPointer()) {
+    bool New = PointerFieldsToUpdate[N].insert(Field.Type().get()).second;
     revng_assert(New);
     revng_log(Log,
               "Wrapped pointer node "
-                << N->ID << " inside wrapper struct with Model ID: "
-                << Struct.ID() << " at address " << &FieldIt->Type());
-    if (Log.isEnabled()) {
-      std::string S;
-      llvm::raw_string_ostream OS{ S };
-      serialize(OS, FieldIt->Type());
-      OS.flush();
-      revng_log(Log, "Pointer: " << S);
-    }
+                << N->ID << " inside wrapper struct (ID: " << Struct.ID()
+                << "):" << serializeToString(Field.Type()));
   }
 
-  // If WrapperSize == 0ULL we use the size of T to set the size of the
-  // generated wrapper struct, otherwise we take WrapperSize, but in that case
-  // it must be larger or equal than T.size() + Offset;
-  Struct.Size() = WrapperSize ? WrapperSize : (Offset + *T.size());
+  Struct.Size() = Offset + *Field.Type()->size();
 
-  revng_assert(Struct.Size() and T.size().has_value()
-               and Struct.Size() >= (T.size().value() + Offset));
-  return QualifiedType{ Path, {} };
+  // If `WrapperSize` is explicitly specified by the caller, ensure that it is
+  // not smaller than the sum of the wrapped type's size with its starting
+  // offset.
+  if (WrapperSize) {
+    revng_assert(WrapperSize >= Struct.Size());
+    Struct.Size() = WrapperSize;
+  }
+
+  return std::move(NewType);
```

Comment:
Wait. On the second thought, are you sure? The value comes from `auto [_, NewType] = Binary.make...(...)`. As far as I understand, RVO is not guaranteed for these. And clang is not complaining either, like it usually does on an harmful `std::move`

Reply:
I guess this is not a new conversation. For me personally `std::move` improves readability. Because it clearly signifies: "hey, this value is no more, no need to keep of it".

I don't really care about this case - since it's the last statement anyway, but it does help me in longer functions.

Reply:
Done.